### PR TITLE
[Clang][Interp] Visit `DecompositionDecl` and create a local variable

### PR DESCRIPTION
**Pull request by yronglin**
The following code should be well-formed:
```C++
float decompose_complex(_Complex float cf) {
  static _Complex float scf;
  auto &[sre, sim] = scf;
  // ok, this is references initialized by constant expressions all the way down
  static_assert(&sre == &__real scf);
  static_assert(&sim == &__imag scf);

  auto [re, im] = cf;
  return re*re + im*im;
}

```
We should visit `DecompositionDecl` and create a local variable but not a create a dummy value directly.

### CODE REVIEW

File: clang/lib/AST/Interp/Compiler.cpp [hunk 1]
```diff
@@ -3593,8 +3593,14 @@ VarCreationState Compiler<Emitter>::visitDecl(const VarDecl *VD) {
   if (R.notCreated())
     return R;
 
-  if (R)
-    return true;
+  if (R) {
```
Comment (yronglin) [inline clang/lib/AST/Interp/Compiler.cpp:3596]: This change just follow current ExprConstant's behavior, but it's no value for this issue, should we remove it?

Comment (tbaederr) [reply clang/lib/AST/Interp/Compiler.cpp:3596]: Yes, remove it please.

File: clang/lib/AST/Interp/Compiler.cpp [hunk 2]
```diff
@@ -5234,6 +5240,9 @@ bool Compiler<Emitter>::visitDeclRef(const ValueDecl *D, const Expr *E) {
           return false;
         };
 
+        if (isa<DecompositionDecl>(VD))
```
Comment (tbaederr) [inline clang/lib/AST/Interp/Compiler.cpp:5243]:
```suggestion
       // DecompositionDecls are just proxies for us.
        if (isa<DecompositionDecl>(VD))
```

**Review by tbaederr — APPROVED** [review]
LGTM with the suggested changes. This might not be perfect wrt. revisiting `DecompositionDecl`s, but we can handle problems when they arise.

File: clang/lib/AST/Interp/Compiler.cpp (hunk 1 above)
Comment (yronglin) [reply clang/lib/AST/Interp/Compiler.cpp]: Removed.

**Pull request merged**